compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1129,6 +1145,115 @@ public void visit(BLangTypeInit cIExpr) {
+            resultType = symTable.semanticError;
+            return symTable.semanticError;
+        } else if (matchingLhsMembers.size() == 1) {
+            // We have a correct match.
+            return matchingLhsMembers.get(0).tsymbol.type;
+        } else {
+            // Multiple matches found.
+            dlog.error(cIExpr.pos, DiagnosticCode.AMBIGUOUS_TYPES, lhsUnion);
+            resultType = symTable.semanticError;
+            return symTable.semanticError;
+        }
+    }
+
+    private boolean isArgsMatchesFunction(List<BLangExpression> invocationArguments, BAttachedFunction function) {
+        if (function == null) {
+            return invocationArguments.isEmpty();
+        }
+
+        if (function.symbol.params.isEmpty() && invocationArguments.isEmpty()) {
+            return true;
+        }
+
+        List<BLangExpression> namedArgs = new ArrayList<>();
+        List<BLangExpression> unnamedArgs = new ArrayList<>();
+        for (BLangExpression argument : invocationArguments) {
+            if (argument.getKind() == NodeKind.NAMED_ARGS_EXPR) {
+                namedArgs.add(argument);
+            } else {
+                unnamedArgs.add(argument);
+            }
+        }
+
+        // All named arguments must be present in function as defaultable parameters.
+        if (!matchDefaultableParameters(function, namedArgs)) {
+            return false;
+        }
+
+        // Given arguments are less than required parameters.
+        int requiredParamCount = function.symbol.params.size();
+        if (requiredParamCount > unnamedArgs.size()) {
+            return false;
+        }
+
+        // No rest params, all (unnamed) args must match params.
+        if (function.symbol.restParam == null && requiredParamCount != unnamedArgs.size()) {
+            return false;
+        }
+
+        // Each arguments must be assignable to required parameter at respective position.
+        List<BVarSymbol> params = function.symbol.params;
+        for (int i = 0, paramsSize = params.size(); i < paramsSize; i++) {
+            BVarSymbol param = params.get(i);
+            BLangExpression argument = unnamedArgs.get(i);
+            if (!types.isAssignable(argument.type, param.type)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean matchDefaultableParameters(BAttachedFunction function, List<BLangExpression> namedArgs) {
+        // More named args given than function can accept.
+        if (function.symbol.defaultableParams.size() < namedArgs.size()) {
+            return false;
+        }
+
+        int matchedParamterCount = 0;
+        for (BVarSymbol defaultableParam : function.symbol.defaultableParams) {
+            for (BLangExpression namedArg : namedArgs) {
+                BLangNamedArgsExpression namedArgExpr = (BLangNamedArgsExpression) namedArg;
+                if (namedArgExpr.name.value.equals(defaultableParam.name.value)) {
```

Comment:
fail fast

Reply:
Failing fast (`continue`) at outer nested `if` didn't make the code any better IMO.
I also find this too nested, I tried a hash-join kind of thing and found it's way more complicated for this.

Early exit:
```java
        int matchedParamterCount = 0;
        for (BVarSymbol defaultableParam : function.symbol.defaultableParams) {
            for (BLangNamedArgsExpression namedArg : namedArgs) {
                if (!namedArg.name.value.equals(defaultableParam.name.value)) {
                    continue;
                }
                BType namedArgExprType = checkExpr(namedArg.expr, env);
                if (types.isAssignable(defaultableParam.type, namedArgExprType)) {
                    matchedParamterCount++;
                } else {
                    // Name matched, type mismatched.
                    return false;
                }
            }
        }
```

Reply:
This is better right? one less level.